enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,480 @@
+    https: <int> # Defaults to 443.
+```
+
+The `ingress.policy.expose` option contains lists of hostnames, NIC names and
+IP addresses. MicroShift will translate them to IP addresses and then update
+the `status.loadBalancer` field in the `Service`. Ovnk will pick up this field
+to configure the iptables rules.
+```yaml
+ingress:
+  expose: # Defaults to IPs in the host. Details below.
+    hostnames:
+    - <hostname>
+    interfaces:
+    - <NIC name>
+    ipAddresses:
+    - <IP address>
+```
+
+### Risks and Mitigations
+Some of the features in this enhancement proposal rely on the
+[LoadBalancer controller](loadbalancer-service-support.md).
+Some of the features rely on NetworkPolicy resources, so the CNI needs
+to support that.
+
+Disabling the router requires a MicroShift restart with all the associated
+consequences (apiserver downtime, etc.).
+
+### Drawbacks
+Some of the features depend on a non-agnostic CNI design, which is the
+`LoadBalancer` controller. A different behavior should be expected if using a
+different CNI.
+
+## Design Details
+#### Why LoadBalancer service
+In the current implementation router ports are fixed and can not be configured.
+Back when there was no support for `LoadBalancer` service types, router was
+forced to use a different way of getting exposed. Using ports 80 and 443 meant
+that NodePort service types can not be used. Using host network would also
+bind port 1936 to the host, which is used for internal metrics. The only option
+that was left was using host ports.
+
+Ports 80 and 443 are configured to use `hostPort` option. This will instruct
+CRI to bind container ports 80 and 443 to the host in the form of iptables
+rules, forwarding incoming traffic to the pod's IP address.
+
+The following limitations apply to the current setup:
+* Iptables rules bypass firewalld settings. This means any firewall
+  configuration applying to host ports will be ignored. It is not possible
+  to control incoming traffic to these ports.
+* NetworkPolicy resources do not apply to host ports. These resources are
+  able to control how traffic reaches which endpoints, including external
+  sources.
+* Any pod using this option is unable to scale within the same node.
+* Port numbers are bound to the deployment, therefore any changes require
+  a restart of the pod.
+
+##### Using LoadBalancer service type
+Common ways of exposing services outside of the cluster are using services of
+types NodePort or LoadBalancer. In MicroShift, ports from the service are
+picked up from ovnk (the CNI) and turned into iptables rules to forward them
+to the service IP. In this case, default router should be exposed using a
+LoadBalancer service type.
+
+ When using this kind of service ovnk will configure special iptables rules
+ that will forward traffic to the service IP instead. These rules are also
+ configured to use the node IP as destination address, as per
+ [LoadBalancer support](loadbalancer-service-support.md) feature.
```

Comment:
There are two services after this enhancement, the `LoadBalancer` and the `ClusterIP`. The `ClusterIP` [already exists](https://github.com/openshift/microshift/blob/main/assets/components/openshift-router/service-internal.yaml) and this is not modified. The enhancement is all about exposing the router outside of the cluster, but the internal traffic remains unchanged.
What ovnk does is to configure new iptables rules forwarding external traffic targeting a specific IP and port to the `LoadBalancer` service IP, using the same port. This gets into regular k8s networking then.
I added a non-goal to clarify the internal traffic remains unchanged.

Reply:
I agree that the internal service is worth mentioning too.

What I was trying to understand is what "the service IP" was. 

>  When using this kind of service ovnk will configure special iptables rules that will forward traffic to the service IP instead

I suppose it refers to the IP from the service' status. Ultimately, it will be the IP address specified in the MicroShift configuration (`ingress.expose`).

Reply:
Clarified the service IP and added a goal about not changing internal access to the router.

Reply:
> When using this kind of service ovnk will configure special iptables rules that will forward traffic to the LoadBalancer service network IP. These rules are also configured to use the node IP as destination address, as per [LoadBalancer support](https://github.com/openshift/enhancements/blob/a82f70e86424825bd39e6a531816f0ba2edaa9b2/enhancements/microshift/loadbalancer-service-support.md) feature.

I'm still a little confused. Will NodeIP be present in the service's status (`.status.loadBalanacer`)? According to [OVN Kubernetes chapter](https://github.com/openshift/enhancements/blob/a82f70e86424825bd39e6a531816f0ba2edaa9b2/enhancements/microshift/loadbalancer-service-support.md#ovn-kubernetes) the destination will be inside the overlay network (service's clusterIP).

Reply:
let me show through an example
```

Reply:
I know  we talked about taking out a lot of the very detailed examples you had in an early draft, but if 1-2 examples help clarify things you could add them.

Reply:
Updated the section to clarify how it works and also add a short example.

Reply:
Thanks! The phrasing alone looks clear enough to me now. I don't insist on the example with `oc` and `iptables` commands.